WalletWasabi/WabiSabi/Crypto/CredentialIssuer.cs

```diff
@@ -196,31 +220,56 @@ public ICommitableCredentialsResponse PrepareResponse(CredentialsRequest registr
 			var areProofsValid = ProofSystem.Verify(transcript, statements, registrationRequest.Proofs);
 			if (!areProofsValid)
 			{
+				// Request was invalid, but all serial numbers were unused.
+				// Ensure nullifier set can't be clogged with invalid serial
+				// numbers. Valid serial numbers are only issued in response to
+				// valid requests, which in turn depend on valid ownership
+				// proofs and the banning mechanism, and they actually matter
+				// for double spending, so only they need to be stored in the
+				// nullifier set.
+				foreach (var serialNumber in presentedSerialNumbers)
+				{
+					SerialNumbers.Remove(serialNumber, out _);
```

Comment:
I didn't look into it but most modern languages use something like siphash these days, with randomized keys.

As for the serials - it's actually 4 per request, and there's more overheads associated with this. Regardless, my main concern is contention, since this is a shared data structure that every credential request writes to. To be lock free on the read path the write path (which does use locking) needs to copy things. It might be able to do that concurrently (i haven't read the implementation).

Anyway, at this point we know the request is invalid, so even if some of the serial numbers matched previously valid ones, any new ones are invalid (not covered by a valid MAC). Since we're already doing this crypto check for all requests, we can use it to prune the nullifier set, because keeping invalid ones wouldn't really help (an attacker trying to flood with invalid requests can just randomize the serial numbers to make sure they don't collide).

Anyway, since it is a pretty straightforward way of conserving server resources and keeping things under control I would rather not let this grow in an unbounded way.

Reply:
looking into it more:

1. the key is the `GroupElement` reference type, which would have a significantly higher memory footprint overall
2. `GroupElement.GetHashCode()` returns `Ge.GetHashCode()` which is the [default implementation for value types which uses reflection](https://docs.microsoft.com/en-us/dotnet/api/system.valuetype.gethashcode?view=net-5.0#System_ValueType_GetHashCode). Since that's in affine form, it's canonical, so correct
3. `ConcurrentDictionary` [uses](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L1884-L1904) the regular bucket linked lists based approach  and `IComparer.GetHashCode(TKey)`

I couldn't find the reflection impl of `GetHashCode`, so I'm not sure if it's randomized at process startup. Regardless, we could override it, and perhaps we should for independent reasons, but i think it's better to keep this aggressive pruning behavior regardless.